packages/legend-query-builder/src/components/QueryBuilder.tsx

```diff
@@ -368,10 +371,30 @@ export const QueryBuilder = observer(
       }
       return null;
     };
+
+    const undo = (): void => {
+      queryBuilderState.changeHistoryState.undo();
+    };
+
+    const redo = (): void => {
+      queryBuilderState.changeHistoryState.redo();
+    };
+
+    useEffect(() => {
+      // this condition is for passing all exisitng tests because when we initialize a queryBuilderState for a test,
+      // we use an empty RawLambda with an empty class and this useEffect is called earlier than initializeWithQuery()
+      if (queryBuilderState.isQuerySupported && queryBuilderState.class) {
```

Comment:
`cacheNewQuery` runs `hashCode` each time. 

Have you tested the performance against larger queries (lots of columns) to ensure the UI isnt affected by this computation between each action?